runtime/src/multi_sig.rs

```diff
@@ -0,0 +1,413 @@
+            if approvals >= approvals_needed {
+                let who_key = Key::try_from(multi_sig.clone().encode())?;
+                match <identity::Module<T>>::get_identity(&who_key) {
+                    Some(id) => {
+                        <identity::CurrentDid>::put(id);
+                    }
+                    _ => return Err("Multisig identity missing"),
+                };
+                let res =
+                    match proposal.dispatch(system::RawOrigin::Signed(multi_sig.clone()).into()) {
+                        Ok(_) => true,
+                        Err(e) => {
+                            let e: DispatchError = e.into();
+                            sr_primitives::print(e);
+                            false
+                        }
+                    };
+                Self::deposit_event(RawEvent::ProposalExecuted(multi_sig, proposal_id, res));
+                return Ok(());
+            } else {
+                return Ok(());
+            }
+        } else {
+            return Err("Invalid proposal");
+        }
+    }
+
+    /// Charges appropriate fee for the proposal
+    fn charge_fee(_multi_sig: T::AccountId, _weight: Weight) -> Result {
+        // TODO use this weight to charge appropriate fee
+        Ok(())
+    }
+
+    /// Accept and process addition of a signer to a multisig
+    pub fn _accept_multi_sig_signer(signer: Signer, auth_id: u64) -> Result {
+        ensure!(
+            <identity::Authorizations<T>>::exists((signer, auth_id)),
+            AuthorizationError::Invalid.into()
+        );
+
+        let auth = <identity::Module<T>>::authorizations((signer, auth_id));
+
+        ensure!(
+            auth.authorization_data == AuthorizationData::AddMultisigSigner,
+            "Not a multi sig signer auth"
+        );
+
+        let wallet_id;
+        if let Signer::Key(multi_sig_key) = auth.authorized_by {
+            wallet_id = T::AccountId::decode(&mut &multi_sig_key.as_slice()[..])
+                .map_err(|_| "Error in decoding multisig address")?;
+        } else {
+            return Err("Error in decoding multisig address");
+        }
+
+        ensure!(
+            <MultiSigSignsRequired<T>>::exists(&wallet_id),
+            "Multi sig does not exist"
+        );
+
+        let wallet_signer = Signer::from(Key::try_from(wallet_id.encode())?);
+        <identity::Module<T>>::consume_auth(wallet_signer, signer, auth_id)?;
+
+        <MultiSigSigners<T>>::insert((wallet_id.clone(), signer), true);
+
+        Self::deposit_event(RawEvent::MultiSigSignerAdded(wallet_id, signer));
+
+        Ok(())
+    }
+
+    pub fn get_next_multi_sig_address(sender: T::AccountId) -> T::AccountId {
+        let nonce: u64 = Self::ms_nonce();
+        let new_nonce: u64 = nonce + 1u64;
+        let h: T::Hash = T::Hashing::hash(&(b"MULTI_SIG", new_nonce, sender.clone()).encode());
+        T::AccountId::decode(&mut &h.encode()[..]).unwrap_or_default()
```

Comment:
I have the same reservation about `decode(encode(_)` as above. In addition, the decode error, unlikely as it may be, would result in the default address, which is undesirable.

Reply:
I hear you but unfortunately, this is the best way. By `decode(encode(_)`, we are basically just changing the type. If rust wasn't strictly typed, we could have gotten away without it.

Reply:
> If rust wasn't strictly typed, we could have gotten away without it.

This is an excellent point. Why not give a multisig account ID its own type?

Reply:
Because we want to use the multisig account id interchangeably with other AccountIds that are used in other modules. The aim is that once a multisig id is generated, it becomes indistinguishable from any other AccountId so that it can do everything that a normal AccountId can do.